tests/ui/option_option.rs

```diff
@@ -0,0 +1,46 @@
+fn input(_: Option<Option<u8>>) {
+}
+
+fn output() -> Option<Option<u8>> {
+    None
+}
+
+fn output_nested() -> Vec<Option<Option<u8>>> {
+    vec![None]
+}
+
+// The lint only generates one warning for this
+fn output_nested_nested() -> Option<Option<Option<u8>>> {
+    None
+}
+
+struct Struct {
+    x: Option<Option<u8>>,
+}
+
+enum Enum {
+    Tuple(Option<Option<u8>>),
+    Struct{x: Option<Option<u8>>},
+}
+
+// The lint allows this
+type OptionOption = Option<Option<u32>>;
+
+// The lint allows this
+fn output_type_alias() -> OptionOption {
+    None
+}
+
+fn main() {
+    input(None);
+    output();
+    output_nested();
+
+    // The lint allows this
+    let local: Option<Option<u8>> = None;
```

Comment:
Do you know why this isn't linted?

Reply:
Based on the discussion in https://github.com/rust-lang-nursery/rust-clippy/issues/97. I thought the lint shouldn't apply to local bindings.

Reply:
I looked through the discussion and could not find anything. Can you quote what you mean?

Reply:
https://github.com/rust-lang-nursery/rust-clippy/issues/97#issuecomment-130176237
"It should catch `Option<Option<..>>` types in function signatures and perhaps typedefs."

I took this to mean _only_ function signatures, struct definitions and enum definitions. I might have been incorrect.

In any case, I think it would be strange to allow the type in expressions but not allow naming those expressions local bindings. Maybe we should move this discussion back into the issue if you want to discuss further.

Reply:
Being conservative is fine here.